src/librustc/ty/wf.rs

```diff
@@ -449,6 +450,33 @@ impl<'a, 'gcx, 'tcx> WfPredicates<'a, 'gcx, 'tcx> {
                   .collect()
     }
 
+    fn nominal_trait_obligations(&mut self,
+                                 trait_ref: &ty::TraitRef<'tcx>)
+                                 -> Vec<traits::PredicateObligation<'tcx>>
+    {
+        // Add in a predicate that `Self:Trait` (where `Trait` is the
+        // current trait).
+        let self_trait = if !trait_ref.has_escaping_regions() {
+            Some(trait_ref.to_predicate())
+        } else {
+            None
+        };
```

Comment:
This feels a bit ad-hoc to me. Can you remember why you added this? Also, when does this "has escaping regions" check return false? I would think .. never ?  You could add a `assert!(!trait_ref.has_escaping_regions());` instead (and perhaps prove me wrong).

If I understand correctly, this is saying that, to prove that `trait Foo { .. }` is well-formed, we have to prove that `Self: Foo`? I'm not sure why that would be useful.

Reply:
There are indeed places where `trait_ref` has escaping regions. There is a call `skip_binder()` above (look for `(*)`).

I believe an example that triggers this is here:
https://github.com/rust-lang/rust/blob/3eadd754730923936ca2ce9c2d303fcac9af099f/src/test/compile-fail/issue-35570.rs#L34-L35

If the trait ref does have escaping regions, it would get filtered out below anyway. Except that `trait_ref.to_predicate()` invokes `ty::Binder::dummy`, which has the assertion you mentioned and we get a panic. Using `ty::Binder::bind` is incorrect and messes up our indices.

On a related note, this fn was copied and modified from the above. I should find a way to remove the repetition.

Reply:
> If I understand correctly, this is saying that, to prove that trait Foo { .. } is well-formed, we have to prove that Self: Foo? I'm not sure why that would be useful.

It gets used to check bounds on fn items as well.